### PR TITLE
Allow IP address as node name

### LINKED ISSUE #309: Handle nodes named by IP address
We're running a local build of master, deployed into our local Kubernetes cluster using the stock services and controllers in deploy/kube-config/influxdb . We see the following from heapster stdout:
```
E0603 02:02:24.529117       1 kube.go:95] Skipping host 10.8.201.152 since looking up its IP failed - lookup 10.8.201.152: invalid domain name
E0603 02:02:24.529212       1 kube.go:95] Skipping host 10.8.201.152 since looking up its IP failed - lookup 10.8.201.152: invalid domain name
E0603 02:02:24.529281       1 kube.go:95] Skipping host 10.8.201.153 since looking up its IP failed - lookup 10.8.201.153: invalid domain name
E0603 02:02:24.529288       1 kube.go:95] Skipping host 10.8.201.153 since looking up its IP failed - lookup 10.8.201.153: invalid domain name
E0603 02:02:24.529320       1 kube.go:95] Skipping host 10.8.201.154 since looking up its IP failed - lookup 10.8.201.154: invalid domain name
E0603 02:02:24.529354       1 kube.go:95] Skipping host 10.8.201.154 since looking up its IP failed - lookup 10.8.201.154: invalid domain name
```
This repeats every 10 seconds.

Here's our cluster:
```
$ kubectl --server=10.8.201.152:8080 get nodes
NAME           LABELS                                   STATUS
10.8.201.152   osResourceAllocationPolicy=over-commit   Ready
10.8.201.153   osResourceAllocationPolicy=over-commit   Ready
10.8.201.154   osResourceAllocationPolicy=one-to-one    Ready
```

### PR DESCRIPTION
**Pull request by jfoy**
Fixes #309

### CODE REVIEW

**Comment by jfoy** [comment]
We have a CCLA as whitepages.com .

File: sources/nodes/kube.go [hunk 1]
```diff
@@ -53,12 +52,41 @@ func (self *kubeNodes) recordGoodNodes(nodes []string) {
...
+		}
+	}
+	if hostname == "" {
+		hostname = node.Name
+	}
+	if nodeInfo.InternalIP == "" {
+		if hostname == nodeInfo.PublicIP {
+			// If the only identifier we have for the node is a public IP, then use it;
+			// don't force a DNS lookup
+			nodeInfo.InternalIP = nodeInfo.PublicIP
```
Comment (vishh) [inline sources/nodes/kube.go]: nit: Can you add a V(4) log here that states that we are using the public IP as-is?

**Comment by vishh** [comment]
Thanks for the PR! LGTM except for the one nit.

**Comment by jfoy** [comment]
Stand by -- integration testing shows this didn't solve the issue.

**Comment by jfoy** [comment]
Looks like there's a deeper issue for us here -- node.Status.Addresses is not getting populated. I'm tracing through why this may be.

**Comment by jfoy** [comment]
This issue turned out to be very close to #202, but I think we've addressed the underlying problem that forced reverting that change with #210.

LegacyHostIP is documented as equivalent to PublicIP in the cloud providers. With this change, the algorithm for choosing a node's IP is as follows:
If the node has an InternalIP, use that address;
otherwise, if the node has a hostname that differs from its PublicIP, resolve the hostname and use that address;
otherwise, use the node's PublicIP (which may have been passed as LegacyHostIP.)

This version is working as expected in our cluster.

**Comment by jfoy** [comment]
Looks like I need to add more unit tests to pass the coverage check.

**Comment by jfoy** [comment]
`make test-unit-cov` is passing on my box; what am I missing here?

**Comment by vishh** [comment]
Unit tests are failing. Can you run `make test-unit`?

**Comment by vishh** [comment]
Coverage is not enforced as of now.

**Comment by jfoy** [comment]
Unit tests are green on my system. (Note: I've changed GOOS=linux to GOOS=darwin for local testing.)
```
201345m0:heapster jfoy$ make test-unit
rm -f heapster
rm -f ./extpoints/extpoints.go
rm -f ./deploy/docker/heapster
go get github.com/tools/godep
go get github.com/progrium/go-extpoints
go get golang.org/x/tools/cmd/vet
hooks/check_boilerplate.sh
hooks/check_gofmt.sh
hooks/run_vet.sh
GOOS=darwin GOARCH=amd64 CGO_ENABLED=0 go generate github.com/GoogleCloudPlatform/heapster
extpoints: Processing file extpoints/interfaces.go
extpoints: Found interfaces: []string{"SourceFactory", "SinkFactory"}
extpoints: Writing file extpoints/extpoints.go
GOOS=darwin GOARCH=amd64 CGO_ENABLED=0 godep go build -a github.com/GoogleCloudPlatform/heapster/...
GOOS=darwin GOARCH=amd64 CGO_ENABLED=0 godep go build -a github.com/GoogleCloudPlatform/heapster
GOOS=darwin GOARCH=amd64 CGO_ENABLED=0 godep go test --test.short github.com/GoogleCloudPlatform/heapster/...
?   	github.com/GoogleCloudPlatform/heapster	[no test files]
?   	github.com/GoogleCloudPlatform/heapster/clusters/coreos	[no test files]
?   	github.com/GoogleCloudPlatform/heapster/extpoints	[no test files]
ok  	github.com/GoogleCloudPlatform/heapster/integration	0.015s
?   	github.com/GoogleCloudPlatform/heapster/manager	[no test files]
?   	github.com/GoogleCloudPlatform/heapster/sinks	[no test files]
ok  	github.com/GoogleCloudPlatform/heapster/sinks/api	0.063s
ok  	github.com/GoogleCloudPlatform/heapster/sinks/gcl	0.012s
?   	github.com/GoogleCloudPlatform/heapster/sinks/gcm	[no test files]
ok  	github.com/GoogleCloudPlatform/heapster/sinks/influxdb	0.011s
ok  	github.com/GoogleCloudPlatform/heapster/sources	0.014s
?   	github.com/GoogleCloudPlatform/heapster/sources/api	[no test files]
ok  	github.com/GoogleCloudPlatform/heapster/sources/datasource	1.321s
ok  	github.com/GoogleCloudPlatform/heapster/sources/nodes	0.014s
?   	github.com/GoogleCloudPlatform/heapster/util/gce	[no test files]
ok  	github.com/GoogleCloudPlatform/heapster/util/gcstore	0.015s
?   	github.com/GoogleCloudPlatform/heapster/validate	[no test files]
?   	github.com/GoogleCloudPlatform/heapster/version	[no test files]
201345m0:heapster jfoy$ echo $?
0
```

**Comment by vishh** [comment]
@cadvisorJenkinsBot: ok to test

**Comment by jfoy** [comment]
The failing test appears to get non-deterministic input:
```
func TestFuzzInput(t *testing.T) {
    var input source_api.AggregateData
    fuzz.New().Fuzz(&input)
    timeseries, err := NewDecoder().Timeseries(input)
    assert.NoError(t, err)
    assert.NotEmpty(t, timeseries)
}
```
Is there a way to get the tests to show a seed, to reproduce a particular fuzzed input?

**Comment by jfoy** [comment]
The failed integration builds appear to be an infrastructure error: 
```
 failed to load docker image "heapster:e2e_test" using temp file "/tmp/kubernetes-minion-71yx171296728" on host "kubernetes-minion-71yx" ("exit status 1") - 
"Warning: Permanently added '104.197.24.127' (ECDSA) to the list of known hosts.
time=\"2015-06-04T19:01:01Z\"
level=\"fatal\"
msg=\"Error: Untar fork/exec /usr/bin/docker: argument list too long\"
ERROR: (gcloud.compute.ssh) [/usr/bin/ssh] exited with return code [1]."
```

**Comment by vishh** [comment]
LGTM.

**Comment by jfoy** [comment]
Thanks!

**Pull request merged**